oeps/oep-0006.rst

```diff
@@ -0,0 +1,410 @@
+
+New BlockScope value: CONTEXT
+-----------------------------
+
+To support context-scoped fields, ``BlockScope`` will allow a new value:
+``CONTEXT``. This new ``BlockScope`` will denote data that is related to all
+instances of this type of XBlock within a particular context.
+
+A "context" is defined by the runtime but is usually a "course" or the nearest
+equivalent. In the Open edX platform, the context may be a course or a content
+library. Different runs of the same course would be considered different
+contexts and would not share data, though UserScope.NONE fields would be copied
+when creating a new course run. On other platforms, a "context" could be defined
+as a blog post or a textbook.
+
+New named scopes
+----------------
+
+As is the situtation today, XBlock runtimes would generally not be expected to
+implement every possible combination of ``BlockScope`` and ``UserScope``.
+Instead, XBlock runtimes are encouraged to implement these new named scopes,
+which would be defined in the XBlock python package:
+
+* ``Scope.context_settings`` (UserScope.NONE, BlockScope.CONTEXT)
+
+  * Example: For an LTI XBlock, a list of the LTI provider URLs and secrets used
+    in the course
+  * Example: For a discussions XBlock: Enable or disable discussions in this
+    course, as well as settings like "allow anonymous discussion posts", "allow
+    anonymous discussion posts to peers", "discussion blackout dates", etc.
+
+* ``Scope.context_preferences`` (UserScope.ONE, BlockScope.CONTEXT)
+
+  * To avoid performance issues, runtimes are not required to support field
+    values larger than 4 KiB for fields that use this scope. This restriction is
+    intended to avoid the temptation to use this type of XBlock field for
+    logging or message passing. Such uses could cause data loss or performance
+    concerns, because the XBlock API does not support appends nor locking for
+    field data.
+  * Example: Enable or disable notifications about new discussion posts in this
+    course
+  * Example: Video playback speed for this course
+  * Example: An XBlock in which students "accumulate building blocks over the
+    course of the semester" (`PLAT-325`_)
+
+.. _PLAT-325: https://openedx.atlassian.net/browse/PLAT-325
+
+New XBlock view: context_settings_view
+--------------------------------------
+
+An XBlock which uses ``Scope.context_settings`` fields will need a way for
+content authors to edit the data in those fields. (e.g. when creating a course,
+the course author may wish to change course-wide settings for the discussion
+XBlock used throughout that course.)
+
+A new XBlock view, ``context_settings_view`` shall be used for this purpose.
+Just like the existing `studio_view` defined by edX Studio, this new view will
+allow an XBlock to return an arbitrary HTML Fragment which contains a form or
+other UI for editing its ``Scope.context_settings`` fields.
+
+It is expected (but not required) that runtimes like edX Studio which want to
```

Comment:
@bradenmacdonald: Does the runtime know the difference between an XBlock instance and the XBlock definition (or class or whatever the appropriate name is)?  Are all existing views (like `studio_view`) XBlock instance views?  

Are we certain that an XBlock view like `context_settings_view` is the right solution for this given these differences?  Should this also be marked as provisional?